recipes/gmail.tsx

```diff
@@ -174,24 +174,196 @@ const updateLimit = handler({
+  // How many times the client will retry after an HTTP failure
+  retries?: number;
+  // In milliseconds, the delay between making any subsequent requests due to failure.
+  delay?: number;
+  // In milliseconds, the amount to permanently increment to the `delay` on every 429 response.
+  delayIncrement?: number;
+}
+
+class GmailClient {
+  private auth: Cell<Auth>;
+  private retries: number;
+  private delay: number;
+  private delayIncrement: number;
+
+  constructor(
+    auth: Cell<Auth>,
+    { retries = 3, delay = 1000, delayIncrement = 100 }: GmailClientConfig = {},
+  ) {
+    this.auth = auth;
+    this.retries = retries;
+    this.delay = delay;
+    this.delayIncrement = delayIncrement;
+  }
+
+  private async refreshAuth() {
+    const body = {
+      refreshToken: this.auth.get().refreshToken,
+    };
+
+    console.log("refreshAuthToken", body);
+
+    const res = await fetch(
+      new URL("/api/integrations/google-oauth/refresh", env.apiUrl),
+      {
+        method: "POST",
+        body: JSON.stringify(body),
+      },
+    );
+    if (!res.ok) {
+      throw new Error("Could not acquired a refresh token.");
+    }
+    const json = await res.json();
+    const authData = json.tokenInfo as Auth;
+    this.auth.update(authData);
+  }
+
+  async fetchEmail(
+    maxResults: number = 100,
+    gmailFilterQuery: string = "in:INBOX",
+  ): Promise<any[]> {
+    const url = new URL(
+      `https://gmail.googleapis.com/gmail/v1/users/me/messages?q=${
+        encodeURIComponent(gmailFilterQuery)
+      }&maxResults=${maxResults}`,
+    );
+
+    const res = await this.googleRequest(url);
+    const json = await res.json();
+    if (
+      !json || !("messages" in json) || !Array.isArray(json.messages)
+    ) {
+      console.log(`No messages found in response: ${JSON.stringify(json)}`);
+      return [];
+    }
+    return json.messages;
+  }
+
+  async fetchBatch(
+    messages: { id: string }[],
+  ): Promise<any[]> {
+    const boundary = `batch_${Math.random().toString(36).substring(2)}`;
+    console.log("Processing batch with boundary", boundary);
+
+    const batchBody = messages.map((message, index) => `
+--${boundary}
+Content-Type: application/http
+Content-ID: <batch-${index}+${message.id}>
+
+GET /gmail/v1/users/me/messages/${message.id}?format=full
+Authorization: Bearer $PLACEHOLDER
+Accept: application/json
+
+`).join("") + `--${boundary}--`;
+
+    console.log("Sending batch request for", messages.length, "messages");
+
+    const batchResponse = await this.googleRequest(
+      new URL(
+        "https://gmail.googleapis.com/batch/gmail/v1",
+      ),
+      {
+        method: "POST",
+        headers: {
+          "Content-Type": `multipart/mixed; boundary=${boundary}`,
+        },
+        body: batchBody,
+      },
+    );
+
+    const responseText = await batchResponse.text();
+    console.log("Received batch response of length:", responseText.length);
+
+    const parts = responseText.split(`--batch_`)
```

Comment:
the response inside the --batch_ can include a 429 as well!  Does this just eat those?

https://developers.google.com/workspace/gmail/api/guides/batch#batch-example-response

Reply:
I can't find explicit docs on the matter, but if 429's result in a 200 HTTP response with JSON containing the error code, I'd have to hope that only 200 responses are represented as text

Reply:
that was my original assumption...

200 in the outside just means "I did the work of fetching each of the sub items and putting it together"

inside each you get an http status code/...

```
--batch_foobarbaz
Content-Type: application/http
Content-ID: <response-item2:12930812@barnyard.example.com>

HTTP/1.1 200 OK
Content-Type: application/json
Content-Length: response_part_2_content_length
ETag: "etag/sheep"

{
  "kind": "farm#animal",
  "etag": "etag/sheep",
  "selfLink": "/farm/v1/animals/sheep",
  "animalName": "sheep",
  "animalAge": 5,
  "peltColor": "green"
}

--batch_foobarbaz
Content-Type: application/http
Content-ID: <response-item3:12930812@barnyard.example.com>

HTTP/1.1 304 Not Modified
ETag: "etag/animals"

--batch_foobarbaz--
```

Reply:
there is a content type / id / length / ... and http version + status code

Reply:
While I don't think this can occur for failure codes, this should handle it

Reply:
Nope it definitely occurs